sdk/storage/src/common/crypt.cpp

```diff
@@ -0,0 +1,188 @@
+        }
+        DWORD objectLength = 0;
+        DWORD dataLength = 0;
+        status = BCryptGetProperty(
+            Handle,
+            BCRYPT_OBJECT_LENGTH,
+            (PBYTE)&objectLength,
+            sizeof(objectLength),
+            &dataLength,
+            0);
+        if (!BCRYPT_SUCCESS(status))
+        {
+          throw std::runtime_error("failed to call bcrypt function");
+        }
+        ContextSize = objectLength;
+        DWORD hashLength = 0;
+        status = BCryptGetProperty(
+            Handle, BCRYPT_HASH_LENGTH, (PBYTE)&hashLength, sizeof(hashLength), &dataLength, 0);
+        if (!BCRYPT_SUCCESS(status))
+        {
+          throw std::runtime_error("failed to call bcrypt function");
+        }
+        HashLength = hashLength;
+      }
+
+      ~AlgorithmProviderInstance() { BCryptCloseAlgorithmProvider(Handle, 0); }
+    };
+
+    static AlgorithmProviderInstance AlgorithmProvider;
+
+    std::string context;
+    context.resize(AlgorithmProvider.ContextSize);
+
+    BCRYPT_HASH_HANDLE hashHandle;
+    NTSTATUS status = BCryptCreateHash(
+        AlgorithmProvider.Handle,
+        &hashHandle,
+        (PUCHAR)context.data(),
+        (ULONG)context.size(),
+        (PUCHAR)key.data(),
+        (ULONG)key.length(),
+        0);
+    if (!BCRYPT_SUCCESS(status))
+    {
+      throw std::runtime_error("failed to call bcrypt function");
+    }
+
+    status = BCryptHashData(hashHandle, (PBYTE)text.data(), (ULONG)text.length(), 0);
+    if (!BCRYPT_SUCCESS(status))
+    {
+      throw std::runtime_error("failed to call bcrypt function");
+    }
+
+    std::string hash;
+    hash.resize(AlgorithmProvider.HashLength);
+    status = BCryptFinishHash(hashHandle, (PUCHAR)hash.data(), (ULONG)hash.length(), 0);
+    if (!BCRYPT_SUCCESS(status))
+    {
+      throw std::runtime_error("failed to call bcrypt function");
+    }
+
+    BCryptDestroyHash(hashHandle);
+
+    return hash;
+  }
+
+  std::string Base64Encode(const std::string& text)
+  {
+    std::string encoded;
+    DWORD encodedLength = DWORD((text.length() + 2) / 3 * 4);
+    encoded.resize(encodedLength);
+
+    CryptBinaryToStringA(
+        (BYTE*)text.data(),
+        (DWORD)text.length(),
+        CRYPT_STRING_BASE64 | CRYPT_STRING_NOCRLF,
+        (LPSTR)encoded.data(),
+        (DWORD*)&encodedLength);
+
+    return encoded;
+  }
+
+  std::string Base64Decode(const std::string& text)
+  {
+    std::string decoded;
+    DWORD decodedLength = DWORD(text.length() / 4 * 3);
+    decoded.resize(decodedLength);
+
+    CryptStringToBinaryA(
+        text.data(),
+        (DWORD)text.length(),
+        CRYPT_STRING_BASE64 | CRYPT_STRING_STRICT,
+        (BYTE*)decoded.data(),
+        &decodedLength,
+        NULL,
+        NULL);
+    decoded.resize(decodedLength);
+    return decoded;
+  }
+
+#else
+
+  std::string HMAC_SHA256(const std::string& text, const std::string& key)
+  {
+    char hash[EVP_MAX_MD_SIZE];
```

Comment:
>char hash[EVP_MAX_MD_SIZE] [](start = 4, length = 26)

Is it going to be dangerous in size given it is allocated on stack?

Reply:
nope, it's 64 bytes. just like you defined 8 `uint64_t`s.